src/CloudNative.CloudEvents.SystemTextJson/JsonEventFormatter.cs

```diff
@@ -0,0 +1,489 @@
+            {
+                throw new ArgumentException($"Structured mode property '{DataBase64PropertyName}' must be a string, when present.");
+            }
+            cloudEvent.Data = dataBase64Element.GetBytesFromBase64();
+        }
+
+        /// <summary>
+        /// Decodes the "data" property provided within a structured-mode message,
+        /// populating the <see cref="CloudEvent.Data"/> property accordingly.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// This implementation converts JSON string tokens to strings when the content type suggests
+        /// that's appropriate, but otherwise returns the token directly.
+        /// </para>
+        /// <para>
+        /// Override this method to provide more specialized conversions.
+        /// </para>
+        /// </remarks>
+        /// <param name="dataElement">The "data" property value within the structured-mode message. Will not be null, and will
+        /// not have a null token type.</param>
+        /// <param name="cloudEvent">The event being decoded. This should not be modified except to
+        /// populate the <see cref="CloudEvent.Data"/> property, but may be used to provide extra
+        /// information such as the data content type. Will not be null.</param>
+        /// <returns>The data to populate in the <see cref="CloudEvent.Data"/> property.</returns>
+        protected virtual void DecodeStructuredModeDataProperty(JsonElement dataElement, CloudEvent cloudEvent) =>
+            cloudEvent.Data = dataElement.ValueKind == JsonValueKind.String && cloudEvent.DataContentType?.StartsWith("text/") == true
+                ? dataElement.GetString()
+                : (object) dataElement.Clone(); // Deliberately cast to object to provide the conditional operator expression type.
+
+        public override byte[] EncodeStructuredModeMessage(CloudEvent cloudEvent, out ContentType contentType)
+        {
+            cloudEvent = cloudEvent ?? throw new ArgumentNullException(nameof(cloudEvent));
+            cloudEvent.ValidateForConversion(nameof(cloudEvent));
+
+            contentType = new ContentType("application/cloudevents+json")
+            {
+                CharSet = Encoding.UTF8.WebName
+            };
+
+            var stream = new MemoryStream();
+            var writer = new Utf8JsonWriter(stream);
+            writer.WriteStartObject();
+            writer.WritePropertyName(CloudEventsSpecVersion.SpecVersionAttribute.Name);
+            writer.WriteStringValue(cloudEvent.SpecVersion.VersionId);
+            var attributes = cloudEvent.GetPopulatedAttributes();
+            foreach (var keyValuePair in attributes)
+            {
+                var attribute = keyValuePair.Key;
+                var value = keyValuePair.Value;
+                writer.WritePropertyName(attribute.Name);
+                // TODO: Maybe we should have an enum associated with CloudEventsAttributeType?
```

Comment:
is this TODO still relevant?

Reply:
Yes - it would be nice to be able to switch on the attribute type, here and in ValidateTokenTypeForAttribute. I've got a separate branch which does that already... I'll create a separate PR to see what you think of it. (I'm not entirely sure yet, myself...)

Reply:
PR with that in is https://github.com/cloudevents/sdk-csharp/pull/112